ddl/column_type_change_test.go

```diff
@@ -0,0 +1,303 @@
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (c1 int, c2 int)")
+	tk.MustExec("insert into t(c1, c2) values (1, 1)")
+	// Enable column change variable.
+	tk.Se.GetSessionVars().EnableChangeColumnType = true
+	defer func() {
+		tk.Se.GetSessionVars().EnableChangeColumnType = false
+	}()
+
+	// use new session to check meta in callback function.
+	internalTK := testkit.NewTestKit(c, s.store)
+	internalTK.MustExec("use test")
+
+	tbl := testGetTableByName(c, tk.Se, "test", "t")
+	c.Assert(tbl, NotNil)
+	c.Assert(len(tbl.Cols()), Equals, 2)
+	c.Assert(getModifyColumn(c, tk.Se.(sessionctx.Context), "test", "t", "c2"), NotNil)
+
+	originalHook := s.dom.DDL().GetHook()
+	defer s.dom.DDL().(ddl.DDLForTest).SetHook(originalHook)
+
+	hook := &ddl.TestDDLCallback{}
+	var checkErr error
+	hook.OnJobRunBeforeExported = func(job *model.Job) {
+		if tbl.Meta().ID != job.TableID {
+			return
+		}
+		switch job.SchemaState {
+		case model.StateNone:
+			tbl = testGetTableByName(c, internalTK.Se, "test", "t")
+			if tbl == nil {
+				checkErr = errors.New("tbl is nil")
+			}
+			if len(tbl.Cols()) != 2 {
+				checkErr = errors.New("len(cols) is not right")
+			}
+		case model.StateDeleteOnly, model.StateWriteOnly, model.StateWriteReorganization:
+			tbl = testGetTableByName(c, internalTK.Se, "test", "t")
+			if tbl == nil {
+				checkErr = errors.New("tbl is nil")
+			}
+			// changingCols has been added into meta.
+			if len(tbl.(*tables.TableCommon).Columns) != 3 {
+				checkErr = errors.New("len(cols) is not right")
+			}
+			if getModifyColumnInAllCols(c, internalTK.Se.(sessionctx.Context), "test", "t", "c2").Flag&parser_mysql.PreventNullInsertFlag == uint(0) {
+				checkErr = errors.New("old col's flag is not right")
+			}
+			if getModifyColumnInAllCols(c, internalTK.Se.(sessionctx.Context), "test", "t", "_Col$_c2") == nil {
+				checkErr = errors.New("changingCol is nil")
+			}
+		}
+	}
+	s.dom.DDL().(ddl.DDLForTest).SetHook(hook)
+	// Alter sql will modify column c2 to tinyint not null.
+	SQL := "alter table t modify column c2 tinyint not null"
+	_, err := tk.Exec(SQL)
+	c.Assert(err, IsNil)
+	// Assert the checkErr in the job of every state.
+	c.Assert(checkErr, IsNil)
+
+	// Check the col meta after the column type change.
+	tbl = testGetTableByName(c, tk.Se, "test", "t")
+	c.Assert(tbl, NotNil)
+	c.Assert(len(tbl.Cols()), Equals, 2)
+	col := getModifyColumn(c, tk.Se.(sessionctx.Context), "test", "t", "c2")
+	c.Assert(col, NotNil)
+	c.Assert(col.Flag&parser_mysql.NotNullFlag, Not(Equals), uint(0))
+	c.Assert(col.Flag&parser_mysql.NoDefaultValueFlag, Not(Equals), uint(0))
+	c.Assert(col.Tp, Equals, parser_mysql.TypeTiny)
+	c.Assert(col.ChangeStateInfo, IsNil)
+	tk.MustQuery("select * from t").Check(testkit.Rows("1 1"))
+}
+
+func getModifyColumnInAllCols(c *C, ctx sessionctx.Context, db, tbl, colName string) *table.Column {
```

Comment:
Could we use `getModifyColumn` instead of `getModifyColumnInAllCols`?

Reply:
`getModifyColumnInAllCols` is used to get invisible column such as `write reorg state` column in ddl callback function.

Reply:
We can add an argument of `cols`, then we can use one function.

Reply:
sounds good